# HTTP Request/Response Cycle - Lab

## Introduction 

In this lab, we'll make use of the `requests` module commands and properties seen in the previous lesson, to extract information for a web service called **"Open Notify"** to access NASA's space data. 

## Objectives

You will be able to:

* Explain the HTTP request/response cycle
* List the status codes of responses and their meanings
* Obtain and interpret status codes from responses
* Make HTTP GET and POST requests in python using the `requests` library

## Open Notify 

[Open Notify](http://open-notify.org/)  is an an open source project to provide a simple programming interface for some of NASA’s awesome data. This takes live raw data from NASA's systems and turn them into APIs related to space and spacecraft. We can access the following information from open notify. 

* Current Location of the International Space Station

* Overhead Pass Predictions for the International Space Station

* Number of People in Space
    
### API endpoints

Open Notify has several API endpoints. 
>An endpoint is a server route that is used to retrieve different data from the API. 

For example, the `/comments` endpoint on the Reddit API might retrieve information about comments, whereas the `/users` endpoint might retrieve data about users. To access them, you would add the endpoint to the base url of the API.

For the OpenNotify API, we have following end points 

1. Current Location of the International Space Station `/iss-now.json`
2. Overhead Pass Predictions for the International Space Station `/iss-pass.json`    
3. Number of People in Space `/astros.json`

The `.json` extension simply tells us that the data is being returned in a JSON format.

In this lab, we'll be querying this API to retrieve live data about the International Space Station (ISS). Details on OpenNofity, endpoints, syntax, and the services it offers can be viewed [Here](http://open-notify.org/Open-Notify-API/)

![](images/iss.jpg)

### Current location of International Space Station

The first endpoint we'll look at on Open Notify is the ` iss-now.json` endpoint (current location of international space station). This endpoint gets the current latitude and longitude of the International Space Station.  Perform the following tasks 
* Make a get request to get the latest position of the international space station from the opennotify api's `iss-now` endpoint at http://api.open-notify.org/iss-now.json
* Check the status code of the response
* Interpret the returned code

In [2]:
import requests

In [5]:
# Your Code Here
resp = requests.get('http://api.open-notify.org/iss-now.json')

In [7]:
print(resp.status_code)

200


In [20]:
resp.status_code == requests.codes.ok

True

* Print the contents of the response and identify its current location

In [21]:
# Your Code Here
print(resp.text)

{"timestamp": 1580287356, "iss_position": {"latitude": "8.2029", "longitude": "68.1332"}, "message": "success"}


### Interpret your results using the API
According to GoogleMaps it is located in the Arabian Sea at the moment the get request was created. 

### Check the next pass of International space station for a given location

Let's repeat the above for the second endpoint `iss-pass.json`. This end point is used to query the next pass of the space station on a given location. Let's just run as above and record your observations.

In [22]:
# Your Code Here
resp2 = requests.get('http://api.open-notify.org/iss-pass.json')

In [23]:
# Your comments 
print(resp2.status_code)

400


In [24]:
resp2.status_code == requests.codes.ok

False

So clearly there is something wrong as we had a 400 response. This is how you should always test your responses for validity. 

If we look at the documentation for the Open Notify API, we see that the ISS Pass endpoint requires two parameters.

> The ISS Pass endpoint returns when the ISS will next pass over a given location on earth. In order to compute this, we need to pass the coordinates of the location to the API. We do this by passing two parameters -- latitude and longitude.

We can do this by adding an optional keyword argument, `params`, to our request. In this case, there are two parameters we need to pass:

* `lat` -- The latitude of the location we want.
* `lon` -- The longitude of the location we want.

Perform the following tasks :
* Set parameters to reflect the lat and long of New York  (40.71, -74)
* Send a get request to Open Notify passing in the lat long parameters as k:v pairs in a dictionary
* Check the status code and interpret
* Print the header information and the returned content

In [38]:
# Your Code Here
parameters = {'lat': 40.71, 'lon': -74}

resp3 = requests.get('http://api.open-notify.org/iss-pass.json', params = parameters)

In [39]:
print(resp3.status_code)

200


In [43]:
resp3.status_code == requests.codes.ok

True

In [44]:
print(dict(resp3.headers))
print(resp3.text)

{'Server': 'nginx/1.10.3', 'Date': 'Wed, 29 Jan 2020 09:02:08 GMT', 'Content-Type': 'application/json', 'Content-Length': '519', 'Connection': 'keep-alive', 'Via': '1.1 vegur'}
{
  "message": "success", 
  "request": {
    "altitude": 100, 
    "datetime": 1580286276, 
    "latitude": 40.71, 
    "longitude": -74.0, 
    "passes": 5
  }, 
  "response": [
    {
      "duration": 255, 
      "risetime": 1580321100
    }, 
    {
      "duration": 636, 
      "risetime": 1580326666
    }, 
    {
      "duration": 638, 
      "risetime": 1580332474
    }, 
    {
      "duration": 569, 
      "risetime": 1580338364
    }, 
    {
      "duration": 587, 
      "risetime": 1580344227
    }
  ]
}



### Check the API and interpret your results - when will ISS pass over NEW York next ?

The ISS will pass over New York "datetime": 1580286276

### Finding the number of people in space

Open Notify has one more API endpoint, `/astros.json`. It tells you how many people are currently in space. The format of the responses can be studied [HERE](http://open-notify.org/Open-Notify-API/People-In-Space/).

Read the above documentation and perform the following tasks:

* Get the response from astros.json endpoint
* Count how many people are currently in space
* List the names of people currently in space.

In [53]:
# Your Code Here
resp4 = requests.get('http://api.open-notify.org/astros.json')
data = resp4.json()

In [54]:
print(resp4.status_code)

200


In [55]:
resp4.status_code == requests.codes.ok

True

In [56]:
print(dict(resp4.headers))
print(resp4.text)

{'Server': 'nginx/1.10.3', 'Date': 'Wed, 29 Jan 2020 09:14:20 GMT', 'Content-Type': 'application/json', 'Content-Length': '314', 'Connection': 'keep-alive', 'access-control-allow-origin': '*'}
{"people": [{"name": "Christina Koch", "craft": "ISS"}, {"name": "Alexander Skvortsov", "craft": "ISS"}, {"name": "Luca Parmitano", "craft": "ISS"}, {"name": "Andrew Morgan", "craft": "ISS"}, {"name": "Oleg Skripochka", "craft": "ISS"}, {"name": "Jessica Meir", "craft": "ISS"}], "number": 6, "message": "success"}


In [59]:
print(data['number'])
print(dict(data))

6
{'people': [{'name': 'Christina Koch', 'craft': 'ISS'}, {'name': 'Alexander Skvortsov', 'craft': 'ISS'}, {'name': 'Luca Parmitano', 'craft': 'ISS'}, {'name': 'Andrew Morgan', 'craft': 'ISS'}, {'name': 'Oleg Skripochka', 'craft': 'ISS'}, {'name': 'Jessica Meir', 'craft': 'ISS'}], 'number': 6, 'message': 'success'}


# Interpret the Results - How many people are in space and what are their names 

There are currently 6 people in space and their names are:
- Christina Koch
- Alexander Skvortsov
- Luca Parmitano
- Andrew Morgan
- Oleg Skripochka
- Jessica Meir

## Summary 

In this lesson, we saw how we can use request and response methods to query an Open API. We also saw how to look at the contents returned with the API calls and how to parse them. Next, we'll look at connecting to APIs which are not OPEN, i.e. we would need to pass in some authentication information and filter the results. 